xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,708 @@
+    if (delegate != null) {
+      return delegate.getListenSockets();
+    }
+    return Collections.emptyList();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getServices() {
+    if (delegate != null) {
+      return delegate.getServices();
+    }
+    return Collections.emptyList();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getImmutableServices() {
+    if (delegate != null) {
+      return delegate.getImmutableServices();
+    }
+    return Collections.emptyList();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getMutableServices() {
+    if (delegate != null) {
+      return delegate.getMutableServices();
+    }
+    return Collections.emptyList();
+  }
+
+  // Must run in SynchronizationContext
+  private void startDelegateServer() {
+    if (restartTimer != null && restartTimer.isPending()) {
+      return;
+    }
+    if (isServing) {
+      return;
+    }
+    if (delegate == null || delegate.isShutdown()) {
+      delegate = delegateBuilder.build();
+    }
+    try {
+      delegate.start();
+      listener.onServing();
+      isServing = true;
+      if (!initialStarted) {
+        initialStarted = true;
+        initialStartFuture.set(null);
+      }
+    } catch (IOException e) {
+      logger.log(Level.FINE, "Fail to start delegate server: {0}", e);
+      if (!initialStarted) {
+        initialStarted = true;
+        initialStartFuture.set(e);
+      }
+      restartTimer = syncContext.schedule(
+        new RestartTask(), retryDelayNano, TimeUnit.NANOSECONDS, timeService);
+    }
+  }
+
+  private final class RestartTask implements Runnable {
+    @Override
+    public void run() {
+      startDelegateServer();
+    }
+  }
+
+  private final class DiscoveryState implements LdsResourceWatcher {
+    private final String resourceName;
+    private final Map<String, RouteDiscoveryState> routeDiscoveryStates = new HashMap<>();
+    private final Map<String, ServerRoutingConfig> routingConfigs = new HashMap<>();
+    // Most recently discovered filter chains.
+    private Map<String, FilterChain> filterChains = new HashMap<>();
+    // Most recently discovered default filter chain.
+    @Nullable
+    private FilterChain defaultFilterChain;
+    // Config for the most recently discovered default filter chain.
+    @Nullable
+    private ServerRoutingConfig defaultRoutingConfig;
+    private boolean stopped;
+    private boolean inflight = false;
+
+    private DiscoveryState(String resourceName) {
+      this.resourceName = checkNotNull(resourceName, "resourceName");
+      xdsClient.watchLdsResource(resourceName, this);
+    }
+
+    @Override
+    public void onChanged(final LdsUpdate update) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          if (stopped) {
+            return;
+          }
+          if (update.listener() == null) {
+            // Ignore updates not for servers.
```

Comment:
Would it actually receive an update "for clients" i.e. an ApiListener? I think this should actually be an assertion that `update.listener()` cannot be null.

Reply:
It seems only if the control plane sends it. `ClientXdsClient` won't tell a LdsUpdate is for client or server.
I guess we can consider it as a temporary error and keep waiting? Temporary/transient error we also do not assert.

Reply:
> ClientXdsClient won't tell a LdsUpdate is for client or server.

yes, it does. Look at `ClientXdsClient.handleLdsResponse` code and it calls `processClientSideListener` or `processServerSideListener` since it knows what is what

Reply:
oh I mean `ClientXdsClient` purely relies on proto it received `listener.hasApiListener()` to notify the watchers, it self does not know. E.g. if we are running a xdsServer, and the xds client receives info that has apiListener, then it will just parse API listener and notify watcher in the xdsServer, then xdsServer ended up hitting this line. So, I think this is a transient error and we should ignore and keep waiting.

Reply:
I don't get this. If you look at the code in `ClientXdsClient.handleLdsResponse`

```
        if (listener.hasApiListener()) {
          ldsUpdate = processClientSideListener(
              listener, retainedRdsResources, enableFaultInjection && isResourceV3);
        } else {
          ldsUpdate = processServerSideListener(
              listener, retainedRdsResources, enableFaultInjection && isResourceV3);
        }
```

the code is clearly using ApiListener to decide client vs server processing.

What you are saying is that a Watcher from XdsServer could still register for client-side-listener because of a code bug (or config mistake?). So this can't be a temporary/transient error

Reply:
> What you are saying is that a Watcher from XdsServer could still register for client-side-listener because of a code bug (or config mistake?)

I mean the when XdsServer registering for listener by calling `xdsClient.watchLdsResource(String resourceName, LdsResourceWatcher watcher) `, it didn't say "i register for client-side-listener" or "i register for server-side-listener", it only specifies the resourcename. So if the xdsClient receives resource matching this resource name, it just parses it and notify the subscriber. Even if this is client side listener but the original subscriber is a server it will still notify. So it is not a bug for XdsServer nor a misconfig.  It is a deficiency or by-design that current implementation in xdsClient can not tell whether it is a xdsServer or a xdsClient that is subscribing to LdsResource.

To prevent this from happening, we might add extra logic in the xdsClient to reject such situation, e.g. by looking at the resource name, because server resource name is from the template while client side it is the authority. But still, from the XdsServerWrapper perspective, if there are deficiency/breakage in other part of the system, it should consider it a transient error and ignore it.